### PR TITLE
[SYCL][Doc] Add new free function queries proposal

### PR DESCRIPTION
**Pull request by Pennycook**
The new proposal makes the following changes:

- Expands upon the motivation for the extension, and provides some
  non-normative recommendations for when it should be used.

- Reformats the description of the functions to match the style used
  by ISO C++ and the newest sections of SYCL 2020.

- Moves all functionality into the this_kernel namespace, to more
  closely align with the this_thread namespace in ISO C++.

- Documents outstanding issues regarding undefined behavior.

Signed-off-by: John Pennycook <john.pennycook@intel.com>

### CODE REVIEW

**Review by rarutyun — COMMENTED** [review]
Overall, looks good but I am planning to look at it deeper after the vacation/holidays (if it is not merged). If this PR is urgent, please don't wait for me since I already put my preliminary review

File: sycl/doc/extensions/FreeFunctionQueries/FreeFunctionQueries.asciidoc
Comment (gmlueck) [inline sycl/doc/extensions/FreeFunctionQueries/FreeFunctionQueries.asciidoc:138]:
```suggestion
_Returns_: A `sycl::id` instance representing the current work-item in the
global iteration space.
```

I assume this is what we intend if this is an ND-range kernel?

File: sycl/doc/extensions/FreeFunctionQueries/FreeFunctionQueries.asciidoc
Comment (gmlueck) [inline sycl/doc/extensions/FreeFunctionQueries/FreeFunctionQueries.asciidoc:150]:
```suggestion
_Returns_: A `sycl::item` instance representing the current work-item in the
global iteration space.
```

File: sycl/doc/extensions/FreeFunctionQueries/FreeFunctionQueries.asciidoc
Comment (Pennycook) [reply sycl/doc/extensions/FreeFunctionQueries/FreeFunctionQueries.asciidoc:122]: Good point.  I'll update this -- I think the restrictions should be the same as `get_item()`.

File: sycl/doc/extensions/FreeFunctionQueries/FreeFunctionQueries.asciidoc
Comment (Pennycook) [reply sycl/doc/extensions/FreeFunctionQueries/FreeFunctionQueries.asciidoc:147]: That's a good question, and I don't know what's preferred here.

The SYCL specification only mentions two execution models [here](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_sycl_kernel_execution_model), and says that "Each work-item is identified by a value of type `item<N>`". So, a user might reasonably expect all kernels to support either `item` or `nd_item`.

I can see an argument that an implementation might have more optimization freedom if it can special-case compilation of kernels taking `id` arguments... But it also seems like it might add more implementation complexity.  If we say that it's illegal to access an `item` when the kernel's parameter is an `id`, we'd effectively be forcing implementations to start distinguishing between three types of kernel (instead of only two).

File: sycl/doc/extensions/FreeFunctionQueries/FreeFunctionQueries.asciidoc
Comment (Pennycook) [reply sycl/doc/extensions/FreeFunctionQueries/FreeFunctionQueries.asciidoc:122]: Fixed in d58457f8260b7b82649c65e64703919cb5bfe65f.

File: sycl/doc/extensions/FreeFunctionQueries/FreeFunctionQueries.asciidoc
Comment (gmlueck) [reply sycl/doc/extensions/FreeFunctionQueries/FreeFunctionQueries.asciidoc:150]: I'm not sure either.  I don't think this would require the implementation to be more complex.  It simply allows an implementation to be more complex (and gain some optimization) if it wants to.

Does DPC++ currently optimize the case when the kernel takes an `id` rather than an `item`?

File: sycl/doc/extensions/FreeFunctionQueries/FreeFunctionQueries.asciidoc
Comment (gmlueck) [inline sycl/doc/extensions/FreeFunctionQueries/FreeFunctionQueries.asciidoc:146]: Sorry, I noticed something else.

The `item` class has a second template parameter `with_offset`.  I think the idea is that you can define your kernel to take `item<Dim, false>` to gain some efficiency if you know it won't be invoked with an offset.  (The `offset` parameter to `parallel_for` is deprecated, but still exists in SYCL 2020.)

The default value of `with_offset` is `true`, so your current specification will always return the "offset" version of `item` even if the user requested the "no offset" version.  It seems like we probably don't want this?  I'm not sure if DPC++ optimizes the "no offset" case currently, but this would be something to check.

An option is to define this free function as always returning the "no offset" case, regardless of how the user defined the kernel.  That might make sense since the "offset" parameter is deprecated.

Comment (Pennycook) [reply sycl/doc/extensions/FreeFunctionQueries/FreeFunctionQueries.asciidoc:146]: > An option is to define this free function as always returning the "no offset" case, regardless of how the user defined the kernel. That might make sense since the "offset" parameter is deprecated.

I agree this makes sense.  I'll add a non-normative note explaining that because the `offset` parameter is deprecated, we expect that eventually we'll return `item<Dimensions>` and it will mean "no offset".

Comment (Pennycook) [reply sycl/doc/extensions/FreeFunctionQueries/FreeFunctionQueries.asciidoc]: Added something in 055e1578a2845bf8092071f172c320aae835bf69.

File: sycl/doc/extensions/FreeFunctionQueries/FreeFunctionQueries.asciidoc
Comment (gmlueck) [reply sycl/doc/extensions/FreeFunctionQueries/FreeFunctionQueries.asciidoc:150]: Yes, agreed that an implementation would need to be more complex if it wanted to flag an error.

I assume your conclusion, therefore, is that `get_item()` can be called even from a kernel that is declared to take an `id` parameter?

Comment (Pennycook) [reply sycl/doc/extensions/FreeFunctionQueries/FreeFunctionQueries.asciidoc:150]: Yes.  I think it's less cognitive burden for users (because there are only two execution models to worry about) and simpler for implementers in the long run (because introducing good error behavior later is desirable).

Do you think we need a non-normative note confirming that we're deliberately allowing `get_item()` to be called from a function accepting an `id`?

Comment (gmlueck) [reply sycl/doc/extensions/FreeFunctionQueries/FreeFunctionQueries.asciidoc:150]: I don't think it is necessary, but I am not opposed.

**Review by gmlueck — DISMISSED** [review]

File: sycl/doc/extensions/FreeFunctionQueries/FreeFunctionQueries.asciidoc
Comment (Pennycook) [reply sycl/doc/extensions/FreeFunctionQueries/FreeFunctionQueries.asciidoc:150]: Ok, I'll leave it as it is for now then.  If anybody is interested they can always read this discussion!

**Comment by bader** [comment]
@Pennycook, please, resolve merge conflicts in README.md file.

**Comment by Pennycook** [comment]
@gmlueck, @bader: The diff doesn't look very nice any more, but if I did everything correctly then https://github.com/intel/llvm/pull/5106/commits/de41bf2d9b4d09ccc905e385608aaeec1d9591bc just moves these changes into experimental/ with the new filename, and https://github.com/intel/llvm/pull/5106/commits/05f1f7773e452d50cc6f37edd72d3d7cb26c9229 moves things out into proposed/.

My justification for putting things into proposed/ now is: 1) we've dropped the experimental namespace; and 2) the new API isn't implemented yet.

**Review by gmlueck — DISMISSED** [review]

**Comment by gmlueck** [comment]
Oh wait.  I just noticed that you deleted the existing "sycl/doc/extensions/experimental/SYCL_EXT_ONEAPI_FREE_FUNCTION_QUERIES.asciidoc".  Was that your intention?

**Comment by Pennycook** [comment]
> Oh wait. I just noticed that you deleted the existing "sycl/doc/extensions/experimental/SYCL_EXT_ONEAPI_FREE_FUNCTION_QUERIES.asciidoc". Was that your intention?

I don't think we need to keep the documentation for the old APIs around.  Since they were experimental we can remove them without deprecation, and that's what I expect we'll do once the new API is implemented.

**Comment by Pennycook** [comment]
@gmlueck: As discussed offline, I've restored the experimental extension in https://github.com/intel/llvm/pull/5106/commits/07f0cbaf9a7e59d9be32f189a320efd60733b5cb.

**Comment by bader** [comment]
@Pennycook, is PR description reflects the changes correctly? I'm using it as a commit message, so I'd like to make sure doesn't conflict with recent updates.

**Comment by Pennycook** [comment]
@bader: I've tweaked things slightly to reflect that we're now adding a new proposal, rather than updating the old extension documentation.  I think it can be merged now.

**Pull request merged**
